### PR TITLE
Merged PoisonedByTiberium and DamagedWithoutFoundations into DamagedByTerrain

### LINKED ISSUE #9690: Veins
- [x] should damage vehicles #10813
- [ ] and buildings #11499
- [x] including correct animations #10813

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27543921-veins-should-damage-vehicles-and-buildings-including-correct-animations?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Mailaender**
This allows veins to damage buildings. Closes #9690.

### CODE REVIEW

File: OpenRA.Mods.Common/DamagedByTerrain.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	[Desc("This actor is attacked by the given weapon when he is in the specified terrain type.")]
+	class DamagedByTerrainInfo : UpgradableTraitInfo, IRulesetLoaded, Requires<HealthInfo>
+	{
+		[Desc("The weapon which is used to damage the actor.")]
+		[WeaponReference, FieldLoader.Require] public readonly string Weapon;
+
+		[Desc("Terrain types where the actor will take damage.")]
+		[FieldLoader.Require] public readonly string[] Terrain = { };
+
+		[Desc("Amount of health below which the actor will not receive further damage in percent.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/DamagedByTerrain.cs]: "Percentage health below which the actor will not receive further damage".

File: OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj [hunk 2]
```diff
@@ -80,6 +80,7 @@
     <Compile Include="SpriteLoaders\R8Loader.cs" />
     <Compile Include="Traits\AttackSwallow.cs" />
     <Compile Include="Traits\Buildings\DamagedWithoutFoundation.cs" />
+    <Compile Include="Traits\Buildings\FreeActorWithDelivery.cs" />
```
Comment (GraionDilach) [inline OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj]: Where did this came from?

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 3]
```diff
@@ -763,6 +763,7 @@
     <Compile Include="Traits\Carryable.cs" />
     <Compile Include="Traits\Carryall.cs" />
     <Compile Include="Traits\Buildings\FreeActorWithDelivery.cs" />
+    <Compile Include="DamagedByTerrain.cs" />
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: This should be in the Traits directory.

File: OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj [hunk 4]
```diff
@@ -80,6 +80,7 @@
     <Compile Include="SpriteLoaders\R8Loader.cs" />
     <Compile Include="Traits\AttackSwallow.cs" />
     <Compile Include="Traits\Buildings\DamagedWithoutFoundation.cs" />
```
Comment (reaperrr) [inline OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj]: This line needs to go as well.

File: OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj (hunk 2 above)
Comment (Mailaender) [inline OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj]: Should be fixed now.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: This needs to go after `OpenRA.Mods.Common`.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Removed `OpenRA.Mods.Common` and sorted the `using`s.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	class DamagedByTerrain : UpgradableTrait<DamagedByTerrainInfo>, ITick, ISync, INotifyAddedToWorld
+	{
+		readonly Health health;
+
+		[Sync] int damageTicks;
+		[Sync] int damageThreshold;
+
+		public DamagedByTerrain(DamagedByTerrainInfo info, Actor self) : base(info)
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Swap order of params.

**Comment by Mailaender** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.GameRules;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("This actor is attacked by the given weapon when he is in the specified terrain type.")]
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: s/is attacked by/receives damage from/, s/in/on/

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: I'd replace the `he is in` with just `on`.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 8]
```diff
@@ -185,6 +185,36 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				if (engineVersion < 20160621)
```
Comment (obrakmann) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: needs updating

**Comment by obrakmann** [comment]
Looks good otherwise, :+1:

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 9]
```diff
@@ -185,6 +185,36 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
...
+
+					if (node.Key.Contains("DamagedWithoutFoundation"))
+					{
+						node.Key = node.Key.Replace("DamagedWithoutFoundation", "DamagedByTerrain");
+						if (!node.Key.StartsWith("-"))
+						{
+							node.Value.Nodes.Add(new MiniYamlNode("Weapon", new MiniYaml("weathering")));
+							node.Value.Nodes.Add(new MiniYamlNode("Terrain", new MiniYaml("Rock")));
+							node.Value.Nodes.Add(new MiniYamlNode("DamageThreshold", new MiniYaml("50")));
+							node.Value.Nodes.Add(new MiniYamlNode("StartOnThreshold", new MiniYaml("true")));
```
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Shall I just remove the upgrade rule completely?

**Comment by Mailaender** [comment]
Updated.

**Comment by Mailaender** [comment]
Rebased.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 10]
```diff
@@ -222,6 +222,40 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
...
+								node.Value.Nodes.Add(new MiniYamlNode("Weapon", new MiniYaml("Tiberium")));
+						}
+					}
+
+					if (node.Key.Contains("DamagedWithoutFoundation"))
+					{
+						node.Key = node.Key.Replace("DamagedWithoutFoundation", "DamagedByTerrain");
+						if (!node.Key.StartsWith("-"))
+						{
+							if (!node.Value.Nodes.Any(a => a.Key == "DamageThreshold"))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: This wants to be `Weapon` instead, I think.

Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Yes. Fixed.

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark:

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/f2b8a21cbd3b4ecd3a39b597d5bde455c443ded0...473fe7cdbdf451ff23740d93852d3d8b82227b03)

**Pull request merged**